### PR TITLE
remove vector-of-pointer-to-vector-of-pointer-to-vector in SourceTables

### PR DESCRIPTION
**Pull request by terhorstd**
This PR simplifies many constructs in `SourceTable` by using vectors with actual objects instead of pointers.

### CODE REVIEW

File: nestkernel/source_table.h [hunk 1]
```diff
@@ -64,17 +64,17 @@ class SourceTable
   /**
    * 3D structure storing gids of presynaptic neurons.
    */
-  std::vector< std::vector< std::vector< Source >* >* > sources_;
+  std::vector< std::vector< std::vector< Source > > > sources_;
```
Comment (jakobj) [inline nestkernel/source_table.h]: the reasoning behind using this structure for the vector was to allow the allocator to create the actual objects
1. independent of each other in the memory space of a specific thread (1st dimension) to improve access speed
2. independent of each other according to synid (2nd dimension) to allow vectors for different syn ids to be distributed in memory to improve freeing of memory when deleting the innermost vectors

are both these conditions still fulfilled if we do not use pointers here?

**Comment by heplesser** [comment]
@jakobj I will post an explanation shortly, all is well. Do NOT MERGE this PR, I will soon post a new one combining this with some other improvements.

--

Dr Hans Ekkehard Plesser
Associate Professor

Faculty of Science and Technology
Norwegian University of Life Sciences
PO Box 5003, 1432 Aas, Norway

Phone +47 6723 1560
Email hans.ekkehard.plesser@nmbu.no<mailto:hans.ekkehard.plesser@nmbu.no>
Home http://arken.nmbu.no/~plesser


On 29 Jun 2018, at 17:39, Jakob Jordan <notifications@github.com<mailto:notifications@github.com>> wrote:


@jakobj commented on this pull request.

Thanks @terhorstd<https://github.com/terhorstd>! I've just added one comment that we need to discuss before further reviewing theses changes.

________________________________

In nestkernel/source_table.h<https://github.com/jakobj/nest-simulator/pull/114#discussion_r199200283>:

> @@ -64,17 +64,17 @@ class SourceTable
   /**
    * 3D structure storing gids of presynaptic neurons.
    */
-  std::vector< std::vector< std::vector< Source >* >* > sources_;
+  std::vector< std::vector< std::vector< Source > > > sources_;


the reasoning behind using this structure for the vector was to allow the allocator to create the actual objects

  1.  independent of each other in the memory space of a specific thread (1st dimension) to improve access speed
  2.  independent of each other according to synid (2nd dimension) to allow vectors for different syn ids to be distributed in memory to improve freeing of memory when deleting the innermost vectors

are both these conditions still fulfilled if we do not use pointers here?

**Comment by heplesser** [comment]
@jakobj The reason the pointers are not necessary is that the `vector` object just contains three pointers (in all known an plausible implementations), to the beginning and end of allocated storage, and to the first available free element. `sizeof(vec)` will return 24 on almost all systems for any type of vector element and vector size. The actual memory for objects is allocated on the free store and this allocation happens whenever the vector grows, which happens in parallel context on the proper threads, ensuring memory locality. Adding yet a layer of pointers only adds code complexity and indirections.

**Comment by heplesser** [comment]
All commits in this PR and some more are now integrated in branch https://github.com/heplesser/nest-simulator/tree/slow-grow-integration-maxblock . @terhorstd  I suggest to close this PR and I will shortly open a new one from my branch.

**Comment by jakobj** [comment]
OK, sounds reasonable. So I'm fine with these changes.

**Comment by jakobj** [comment]
addressed by #117, closing

**Pull request closed without merging**